src/releases/mod.rs

```diff
@@ -28,27 +30,66 @@ pub(crate) struct Release {
     pub(crate) package_name: Option<String>,
 }
 
-#[derive(Clone, Debug, Eq, PartialEq)]
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
 pub(crate) struct CurrentVersions {
-    pub(crate) stable: Version,
-    pub(crate) prerelease: Option<Version>,
+    pub(crate) stable: Option<Version>,
+    pub(crate) prereleases: Prereleases,
 }
 
+type Prereleases = BTreeMap<Version, BTreeMap<Label, Prerelease>>;
+
 impl CurrentVersions {
-    pub(crate) fn latest(&self) -> &Version {
-        self.prerelease.as_ref().unwrap_or(&self.stable)
+    pub(crate) fn into_latest(mut self) -> Option<Version> {
+        self.prereleases
+            .pop_last()
+            .map(|(version, mut pres)| {
+                let mut version = version;
+                version.pre = Some(
+                    pres.pop_last()
+                        .expect("This map is never allowed to be empty")
+                        .1,
+                );
+                version
+            })
+            .or(self.stable)
     }
 
-    pub(crate) fn into_latest(self) -> Version {
-        self.prerelease.unwrap_or(self.stable)
+    pub(crate) fn insert_prerelease(&mut self, prerelease: Version) {
+        let (stable_component, pre) = {
+            let mut version = prerelease;
+            let pre = version.pre.take().expect("This is a prerelease");
+            (version, pre)
+        };
+        let recorded_pre = self
+            .prereleases
+            .get(&stable_component)
+            .and_then(|pres| pres.get(&pre.label));
+        if let Some(recorded_pre) = recorded_pre {
+            if recorded_pre >= &pre {
+                return;
+            }
+        }
+        self.prereleases
+            .entry(stable_component)
+            .or_default()
+            .insert(pre.label.clone(), pre);
+    }
+
+    pub(crate) fn replace_stable_if_newer(&mut self, version: Version) {
+        if let Some(stable) = &self.stable {
+            if stable >= &version {
+                return;
+            }
+        }
+        self.stable = Some(version);
```

Comment:
I think I want two types to enforce the invariants since right now there's a bunch of runtime logic which is _hopefully_ correct enforcing that, but not necessarily guaranteed. Rust can do better